contrib: Introduce nitrpg

Morriar opened the pull request:
We have been talking about this a long ago. I made some demos, merged the full Github API stack...

Tonight is the nigth!

In this first PR we introduce some game basics and just enough to show how the game is supposed to work.
You can run it as this but the game only rewards new pull requests and +1 comments.
Further PR will introduce more capabilities like Github history reload, achievements, experience...

The game is separated into two tools:

* `listener` is the Github hook that capture all the github events on a repo.
* `web` is a Nitcorn client that display the game as a HTML website.

Four first commits are from #1141, please disregard.

Morriar commented:
The diff seems big but half the lines are HTML :p

privat commented on contrib/nitrpg/src/game.nit:
```diff
@@ -0,0 +1,270 @@
+	redef fun compare(a, b) do
+		if a.nitcoins < b.nitcoins then return 1
+		if a.nitcoins > b.nitcoins then return -1
+		return 0
```
~~~nit
return b.nitcoins <=> a.nitcoins
~~~

privat commented on contrib/nitrpg/src/game.nit:
```diff
@@ -0,0 +1,270 @@
+# `GithubEvent`.
+#
+# Depending on the received event, the reactor is used to update game data.
+interface GameReactor
```
La doc du module nous dit de regarder ici mais: `GameReactor` n'est pas utilisé dans le reste du module donc on sait pas trop comment s'en servir (qui instancie les sous-classes, où elles sont stockés, qui appelle `react_event`)

privat commented on contrib/nitrpg/src/listener.nit:
```diff
@@ -0,0 +1,57 @@
+	# Dispatch event to registered `reactors`.
+	redef fun apply_event(event) do
+		var game = new Game(api, event.repo)
+		game.verbose_lvl = 1
```
tu bricoles les niveaux de verbosité comme ça toi?

privat commented on contrib/nitrpg/src/reactors.nit:
```diff
@@ -0,0 +1,63 @@
+	redef fun react_player_event(game) do
+		if action == "opened" then
+			var player = pull.user.player(game)
+			player.nitcoins += 10
```
fixme?: remove magic numbers

privat commented on contrib/nitrpg/src/reactors.nit:
```diff
@@ -0,0 +1,63 @@
+	#
+	# TODO only give nitcoins if reviewers < 2
+	redef fun react_player_event(game) do
+		if comment.body.has("\\+1\\b".to_re) then
```
hum...

privat commented on contrib/nitrpg/src/templates/panels.nit:
```diff
@@ -0,0 +1,306 @@
+
+	redef fun render_title do
+		add "<span class=\"glyphicon glyphicon-home\"></span>&nbsp;&nbsp;"
+		add "<a href=\"{game.url}\">{game.name}</a>"
```
escape html for url and name (to avoid issues/injection)?

privat commented on contrib/nitrpg/src/templates/panels.nit:
```diff
@@ -0,0 +1,306 @@
+
+	redef fun render_body do
+		var q = "is:open label:need_review sort:updated-asc " +
+			"-involves:{player.name}"
```
great! inspired by the dashboards?

privat commented on contrib/nitrpg/src/templates/templates_base.nit:
```diff
@@ -0,0 +1,33 @@
+redef class GameEntity
+
+	# URL to this game entity page.
+	fun url: String do return game.url / key
```
add `escaped_url` to factorise html escaping

Morriar replied:
En fait, les keys sont basés sur les noms issues de l'API Github, donc c'est déjà pré-formaté.

privat commented on contrib/nitrpg/src/web.nit:
```diff
@@ -0,0 +1,163 @@
+
+var iface = "{host}:{port}"
+var vh = new VirtualHost(iface)
+vh.routes.add new Route("/styles/", new FileServer("www/styles"))
```
because you will get "http://nitlanguage.org/rpg" you sould be able to configure a prefix or a root something

Morriar replied:
Not needed here because the FileServer render files from the current directory (where `web` is started).

But your remarks apply for all links found in templates, so I will do it :)

privat commented:
+1

privat commented on contrib/nitrpg/Makefile:
```diff
@@ -0,0 +1,26 @@
+all: listener web
+
+listener:
+	nitc src/listener.nit
```
utilise `../../bin/nitc`

ou fait une variable
~~~
NITC=../../bin/nitc
~~~
puis
~~~
$(NITC) src/listener.nit
~~~

Morriar commented:
Demo is online: http://nitlanguage.org/rpg/games/privat/nit

R4PaSs commented:
MRW checking demo
![](https://i.imgflip.com/h7km2.jpg)

R4PaSs commented:
Cool to see it arrive ! You have my +1 sir !

Morriar commented:
Reroll:

* [x] added some fixme and todo following @privat remarks
* [x] added an argument to precise root  url
* [x] fixed Makefile
* [x] fixed DocReactor documentation
* [x] fixed PlayerComparator

privat commented on contrib/nitrpg/.gitignore:
```diff
@@ -0,0 +1,2 @@
+.github_data
+nitrpg_data
```
add produced binaries to have a clean dir after make

privat commented on contrib/nitrpg/Makefile:
```diff
@@ -0,0 +1,28 @@
+	$(NITC) src/web.nit
+
+clean:
+	rm listener web
```
also clean data?

Morriar replied:
You don't want to lost all your saved games don't you?

Morriar commented:
Fixed .gitignore

privat commented:
+1
